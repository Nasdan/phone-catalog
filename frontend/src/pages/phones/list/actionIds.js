export const actionIds =  {
  FETCH_PHONES_SUCCESS: '[Phone List] [1] Available phones fetched succesfully from API',
}
