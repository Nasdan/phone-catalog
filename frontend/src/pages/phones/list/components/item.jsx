import React from 'react';
import PropTypes from 'prop-types';
import { cx } from 'emotion';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Typography from '@material-ui/core/Typography';
import { item, card, cardContent, image, title, price } from './item.styles';

export const Item = props => (
  <li className={cx(item, props.className)}>
    <Card className={card}>
      <CardContent className={cardContent}>
        <div className={image}>
          <img src={props.phone.imageUrl} />
        </div>
        <Typography className={title}>{props.phone.title}</Typography>
        <Typography className={price}>
          <h2>{props.phone.price}</h2>
        </Typography>
      </CardContent>
    </Card>
  </li>
);

Item.propTypes = {
  phone: PropTypes.object.isRequired,
  className: PropTypes.string.isRequired,
};
