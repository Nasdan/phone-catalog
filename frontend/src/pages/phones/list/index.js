export * from './container';
export { phoneListReducer } from './reducer';
