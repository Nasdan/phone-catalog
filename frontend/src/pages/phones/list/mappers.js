export const mapPhonesModelToViewModel = phones =>
  phones.map(mapPhoneModelToViewModel);

const mapPhoneModelToViewModel = phone => ({
  ...phone,
  price: `${phone.price} â‚¬`,
});
