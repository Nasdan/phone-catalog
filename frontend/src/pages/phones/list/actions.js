import { trackPromise } from 'react-promise-tracker';
import { phoneAPI } from '../../../rest-api/api/phone';
import { actionIds } from './actionIds';

export const fetchPhones = () => dispatch =>
  trackPromise(
    phoneAPI
      .fetchPhones()
      .then(phones => {
        dispatch(updatePhones(phones));
      })
      .catch(error => {
        console.error(error);
      })
  );

const updatePhones = phones => ({
  type: actionIds.FETCH_PHONES_SUCCESS,
  payload: phones,
});
