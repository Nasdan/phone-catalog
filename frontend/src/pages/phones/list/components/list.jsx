import React from 'react';
import PropTypes from 'prop-types';
import { Item } from './item';
import { list, item } from './list.styles';

export const List = props => (
  <ul className={list}>
    {props.phones.map(phone => (
      <Item key={phone.id} className={item} phone={phone} />
    ))}
  </ul>
);

List.propTypes = {
  phones: PropTypes.array.isRequired,
};
