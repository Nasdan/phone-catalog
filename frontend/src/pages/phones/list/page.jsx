import React from 'react';
import PropTypes from 'prop-types';
import { Header, List } from './components';

export const PhoneListPage = props => (
  <React.Fragment>
    <Header />
    <List phones={props.phones} />
  </React.Fragment>
);

PhoneListPage.propTypes = {
  phones: PropTypes.array.isRequired,
};
