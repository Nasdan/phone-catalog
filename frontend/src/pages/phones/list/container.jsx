import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { routes } from '../../../common/constants/routes';
import { buildRoute } from '../../../common/helpers/routes';
import { phoneAPI } from '../../../rest-api/api/phone';
import { mapPhonesModelToViewModel } from './mappers';
import { PhoneListPage } from './page';

class InnerPhoneListContainer extends React.PureComponent {
  state = {
    phones: [],
  };

  componentDidMount() {
    this.loadPhones();
  }

  loadPhones = () => {
    phoneAPI.fetchPhones().then(phones => {
      this.setState({
        phones: mapPhonesModelToViewModel(phones),
      });
    });
  };

  handleItemClick = phone => {
    const route = buildRoute(routes.phones.detail, { id: phone.id });
    this.props.history.push(route);
  };

  render() {
    return (
      <PhoneListPage
        phones={this.state.phones}
        onItemClick={this.handleItemClick}
      />
    );
  }
}

InnerPhoneListContainer.propTypes = {
  history: PropTypes.object.isRequired,
};

export const PhoneListContainer = withRouter(InnerPhoneListContainer);
