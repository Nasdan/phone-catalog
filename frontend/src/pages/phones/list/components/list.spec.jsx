import React from 'react';
import { shallow } from 'enzyme';
import { Item } from './item';
import { List } from './list';

describe('pages/phones/list/components/list specs', () => {
  it('should render as expected when feeding props', () => {
    // Arrange
    const props = {
      phones: [
        {
          id: 1,
          title: 'test title',
          imageUrl: 'test imageUrl',
          price: 'test price',
        },
      ],
      onItemClick: jest.fn(),
    };

    // Act
    const component = shallow(<List {...props} />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  it('should call to onClick prop when simulate Item click', () => {
    // Arrange
    const props = {
      phones: [
        {
          id: 1,
          title: 'test title',
          imageUrl: 'test imageUrl',
          price: 'test price',
        },
      ],
      onItemClick: jest.fn(),
    };

    // Act
    const component = shallow(<List {...props} />);

    const itemComponent = component.find(Item);
    itemComponent.simulate('click');

    // Assert
    expect(props.onItemClick).toHaveBeenCalled();
  });
});
