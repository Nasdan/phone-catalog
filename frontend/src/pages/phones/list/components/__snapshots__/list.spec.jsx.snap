// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/phones/list/components/list specs should render as expected when feeding props 1`] = `
<ul
  className="css-r1kb16"
>
  <Item
    className="css-l9u6wf"
    key="1"
    onClick={[MockFunction]}
    phone={
      Object {
        "id": 1,
        "imageUrl": "test imageUrl",
        "price": "test price",
        "title": "test title",
      }
    }
  />
</ul>
`;
